SystemInformer/proctree.c

```diff
@@ -1451,6 +1486,155 @@ static VOID PhpUpdateProcessNodePriorityBoost(
     }
 }
 
+// This method was added to be called from the sort logic for sorting this column
+// It needs to refresh all nodes's (PPH_PROCESS_NODE) service names (in the sorting function referred to generically as node1, node2)
+// This was the logic that existed in the switch case for case PHPRTLC_SERVICENAMES:
+// Then instead of 'node' and 'processItem' utilize 'ProcessNode' and 'ProcessNode->ProcessItem' to referr to the various available attributes
+static VOID PhpUpdateProcessServiceNames(
+    _Inout_ PPH_PROCESS_NODE ProcessNode
+)
+{
+    if (!(ProcessNode->ValidMask & PHPN_SERVICENAMES))		// Bitwise flag for calling this update method
+    {
+        if (!ProcessNode->SvcNames)
+        {
+            ProcessNode->SvcNames = PhCreateString(L"");
+        }
+
+        if ((ProcessNode->ProcessItem->ServiceList && ProcessNode->ProcessItem->ServiceList->Count != 0))
+        {
+            PH_STRING_BUILDER stringBuilder;
+            PhInitializeStringBuilder(&stringBuilder, 1000);
+
+            // Based on itemtips.c	 :	serviceList = PhCreateList(Process->ServiceList->Count);
+            //		I don't think ServiceListLock is needed, since we are not modyfing this, just simply reading it.
+            //		In itemtips.c its inserting items
+
+            //PhAcquireQueuedLockShared(&ProcessNode->ProcessItem->ServiceListLock);
+
+            for (ULONG i = 0; i < ProcessNode->ProcessItem->ServiceList->Count; i++)
+            {
+                PPH_SERVICE_ITEM serviceItem = ProcessNode->ProcessItem->ServiceList->Items[i];
+
+                PhAppendStringBuilder(&stringBuilder, &serviceItem->Name->sr);
+                PhAppendStringBuilder2(&stringBuilder, L" (");
+                PhAppendStringBuilder(&stringBuilder, &serviceItem->DisplayName->sr);
+                PhAppendStringBuilder2(&stringBuilder, L")");
+
+                if (i + 1 < ProcessNode->ProcessItem->ServiceList->Count)
+                {
+                    PhAppendStringBuilder2(&stringBuilder, L"  |  ");
+                }
+            }
+
+            //PhReleaseQueuedLockShared(&ProcessNode->ProcessItem->ServiceListLock);
+
+            // Destroy the blank string above, before assigneding to this instead
+            PhDereferenceObject(ProcessNode->SvcNames);
+
+            ProcessNode->SvcNames = PhFinalStringBuilderString(&stringBuilder);
+        }
+
+        ProcessNode->ValidMask |= PHPN_SERVICENAMES;
+    }
+}
+
+static VOID PhpUpdateProcessServiceCount(
+    _Inout_ PPH_PROCESS_NODE ProcessNode
+)
+{
+    if (!(ProcessNode->ValidMask & PHPN_SVCCOUNT))		// Bitwise flag for calling this update method
+    {
+        ProcessNode->SvcCount = 0;     // Default to nothing
+
+        PhAcquireQueuedLockShared(&ProcessNode->ProcessItem->ServiceListLock);
```

Comment:
I think you should probably do this in one call instead of two. You seem to be walking the list twice. Is there even value in having a "count"?

Reply:
A count of services hosted by a process seems nice to bubble up the ones that contain the most, I do think. But perhaps I’m over thinking it.